core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java

```diff
@@ -35,10 +37,11 @@ public GenericPlanNode getGenericPlanNode() {
 		return top;
 	}
 
-	// node.getParentNode() is not reliable because nodes are reused and parent is not maintained! This is why we use a
-	// queue to maintain the effective parent stack.
 	@Override
 	protected void meetNode(QueryModelNode node) {
+		assert !visited.containsKey(node);
+		visited.put(node, node);
```

Comment:
What was the reason you added this? Did you see any cycles in query plan?

Reply:
This is something _you_ added, I think. I assumed it was necessary, but was this part of the throwaway test code as well?

Reply:
Yeah. It was.

Reply:
I intended for you to just revert my commits when you had figured out how to test the issue properly and how to fix it.

Reply:
I'll take it out. See, this is why I don't like throwaway code. It has a nast habit of sticking around.